vital/algo/serialize_metadata.h

```diff
@@ -0,0 +1,134 @@
+// This file is part of KWIVER, and is distributed under the
+// OSI-approved BSD 3-Clause License. See top-level LICENSE file or
+// https://github.com/Kitware/kwiver/blob/master/LICENSE for details.
+
+/**
+ * \file
+ * \brief Interface for serialize_metadata \link
+ * kwiver::vital::algo::algorithm_def algorithm
+ *        definition \endlink.
+ */
+
+#ifndef VITAL_ALGO_SERIALIZE_METADATA_H_
+#define VITAL_ALGO_SERIALIZE_METADATA_H_
+
+#include <vital/vital_config.h>
+
+#include <string>
+
+#include <vital/algo/algorithm.h>
+#include <vital/algorithm_capabilities.h>
+#include <vital/types/metadata.h>
+#include <vital/types/metadata_map.h>
+
+namespace kwiver {
+
+namespace vital {
+
+namespace algo {
+
+/// An abstract base class for reading and writing meatadata maps
+
+/**
+ * This class represents an abstract interface for reading and writing
+ * video metadata.
+ */
+class VITAL_ALGO_EXPORT serialize_metadata
+  : public kwiver::vital::algorithm_def< serialize_metadata >
```

Comment:
@mleotta might remember better than I do, but I think it was because `vital/algo/data_serializer`, which I believe you're referring to, goes from data to a string representation and vice versa. The goal of this metadata serialization is to explicitly read and write to disk.

Perhaps this indirectly addresses the other two questions. I was initially just going to update this to `serialize_metadata_map` but maybe something like `metadata_map_io` would make more sense. I think it should be possible to switch to taking streams as input but there's a bit of logic for checking that filenames are valid which I think would have to be moved to the calling code if this switched to taking streams. That makes me a little hesitant to switch to streams but if you feel it's important to switch I can.

Reply:
Another option might be to add *overloads* taking streams, and have the overloads that take files do the checking, open the file (in order to get a stream), then hand off to the stream overloads.

Alternatively, maybe we should just have a utility function that tries to return a stream (may need to return it as a `unique_ptr`) from a file.

Reply:
I feel like the overload makes sense in this case. But I do also think the utility function you describe could be broadly useful.

Reply:
The data serializers have very specific requirements along with naming conventions which are not really what you want here. I like the comment about making this `metadata_map_io` to separate it from the serializers and overloading to handle file names and streams seems like a good approach. These type of algorithms/classes get used in unexpected ways if they turn out to be useful.